### PR TITLE
[Core]Adding orderReference support to core

### PR DESCRIPTION
**Pull request by makarid**
#3376

### CODE REVIEW

**Comment by mdvx** [comment]
Should name the field orderUserReference, that would save on confusion.

**Comment by makarid** [comment]
OK

**Comment by makarid** [comment]
I am getting the following Access Denied error on Travis CI:

[ERROR] Plugin org.apache.maven.plugins:maven-source-plugin:3.2.1 or one of its dependencies could not be resolved: Failed to read artifact descriptor for org.apache.maven.plugins:maven-source-plugin:jar:3.2.1: Could not transfer artifact org.apache.maven.plugins:maven-source-plugin:pom:3.2.1 from/to central (https://repo.maven.apache.org/maven2): Access denied to: https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-source-plugin/3.2.1/maven-source-plugin-3.2.1.pom , ReasonPhrase:Forbidden. -> [Help 1]

**Comment by makarid** [comment]
Do i need to update from 3.2.1 to newer or something?

File: xchange-core/src/main/java/org/knowm/xchange/dto/trade/UserTrade.java [hunk 1]
```diff
@@ -74,6 +77,14 @@ public Currency getFeeCurrency() {
     return feeCurrency;
   }
 
+  public String getOrderUserReference() {
+    return orderUserReference;
+  }
+
+  public void setOrderUserReference(String orderUserReference) {
+    this.orderUserReference = orderUserReference;
+  }
```
Comment (walec51) [inline xchange-core/src/main/java/org/knowm/xchange/dto/trade/UserTrade.java]: UserTrade should not be mutable

Comment (walec51) [reply xchange-core/src/main/java/org/knowm/xchange/dto/trade/UserTrade.java]: please also make Trade immutable again as I see that its you who added two setters the a year ago

Comment (makarid) [reply xchange-core/src/main/java/org/knowm/xchange/dto/trade/UserTrade.java]: OK, i will delete the two setters on Trade.

Comment (makarid) [reply xchange-core/src/main/java/org/knowm/xchange/dto/trade/UserTrade.java]: I will completely delete the two maker and taker fields and i will add the orderUserReference to the UserTrade constructor. Is this what you want?

**Comment by makarid** [comment]
I order to not make any mistake and then i have to redo my changes, i am telling you what i will do and tell me if it is OK.

1)Delete the makerOrderId and takerOrderId fields from Trade object.
2)Add a new field on UserTrade.class constructor and replace all new UserTrade constructors from the project with UserTrade.Builder.

Is that OK? I am waiting for your response @walec51 .Thanks

**Comment by walec51** [comment]
1) if those two fields are not used anywhere then yes remove them, but if the are then just make them immutable (no setter, set via constructor / builder)

2) yes, using builders is the way to go

**Comment by makarid** [comment]
> 1. if those two fields are not used anywhere then yes remove them, but if the are then just make them immutable (no setter, set via constructor / builder)
> 2. yes, using builders is the way to go

Done. Please check. Thanks a lot!

**Comment by walec51** [comment]
thank up for this big contribution and sorry it took some time to get merged - that was a lot of files to read :)

**Comment by makarid** [comment]
> thank up for this big contribution and sorry it took some time to get merged - that was a lot of files to read :)

Totally understood! :)

**Pull request merged**